### ISSUE TITLE
Bundle returned when API key doesn't have read access

### ISSUE THREAD

**Issue by timskovjacobsen** [levinsen-software/warehouse-python#1]
Hi

I was wondering why some of the warehouse-client package has "raise for status"-like behavior, while other parts do not.

I recently had the situation below. The API key does not have read access on a bundle, but the bundle still gets returned to the caller:

```python
from pathlib import Path

import warehouse

WAREHOUSE_URL = "https://warehouse.domain.com"
WAREHOUSE_APIKEY = Path("~/.warehouse").expanduser().read_text().strip("\n")

client = warehouse.Client(WAREHOUSE_URL, warehouse.ApikeyAuth(WAREHOUSE_APIKEY))

# A bundle is returned here even though the API lacks read permission
# -> req.status is 403 Forbidden
bundle = client.bundle("<some-id>")

if bundle:
    # Bundle is there, but can't be interacted with
    # -> requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
    r = bundle.get_properties()
else:
    raise ValueError("Bundle not found")
```

I see that there are some error handling checks inside `client.py` and `file.py`, but not in other places in the API.

I would argue that warehouse-client should either raise the error itself, return bundle as `None`, or return some kind of error-handling mechanism like `bundle.status` to the caller so they can check and handle the error.

Currently, the bundle object has these attributes/methods:

```python
(Pdb) [attr for attr in dir(bundle) if not attr.startswith("__")]
[
    'delete',
    'files',
    'findFile',
    'findFiles',
    'find_file',
    'find_files',
    'getProperties',
    'get_properties',
    'id',
    'restore',
    'trash',
    'updateProperties', 
    'update_properties',
    'uploadFile',
    'upload_file',
    'wh'
]
```

**Comment by kennylevinsen** [comment]
Currently, `client.bundle` and similar methods just set up client-side object without any communication in order to save roundtrips in case the script is accessing many bundles (e.g., hundreds or more). As it does does not do any validation, the bundle might not even exist.

While it will cost an additional roundtrip, doing a quick HEAD probe might make sense. We could always add a kwarg to optionally bypass it if one wants to save the roundtrip and delay errors until the bundle is used.

**Comment by timskovjacobsen** [comment]
I understand. It makes sense to delay the communication to limit the number of requests.

It could be done at method level perhaps, like it's done e.g. for `WHFile.trash()`. In that way, one would still get the "empty" bundle back, but at first interaction with it see a fitting error like "403 Forbidden" instead of failing inside `req.json()` with `requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`.

**Comment by filips** [comment]
We just released version 1.0.10 of warehouse-python. It has improved error handling for all the API endpoints that didn't check the HTTP status code prior to attempting `req.json()`.

**Comment by timskovjacobsen** [comment]
Thanks for improving the error handling.
